File: TP/CH03/autocompletion/java/Autocompletion.java
import java.util.regex.*;
import java.io.*;
import java.util.*;
import java.lang.Math;
import java.util.stream.Collectors;

public class Autocompletion {
	private LaplaceBiGram modele = new LaplaceBiGram();
	private List<Tuple<String, String>> eval = new ArrayList<>();

	public void entrainer(String url){
		try {
			BufferedReader f = new BufferedReader(new InputStreamReader(new FileInputStream(url), "UTF-8"));
			List<String[]> data = new ArrayList<>();
			String l;
			while((l = f.readLine()) != null) {
				String[] phrase = l.toLowerCase().split(" ");
				if (phrase.length > 0) data.add(phrase);
			}
			this.modele.entrainer(data);
		} catch(IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public List<Tuple<String, Double>> estimer(String phrase, int nbr){
		List<Tuple<String, Double>> motsScores;
		motsScores = this.modele.estimer(phrase.toLowerCase().split(" "));
		return motsScores.subList(0, nbr);
	}

	public void charger_modele(String url){
		try {
			FileInputStream f = new FileInputStream(url);
			ObjectInputStream in = new ObjectInputStream(f);
			modele = (LaplaceBiGram) in.readObject();
			in.close();
			f.close();
		} catch(IOException|ClassNotFoundException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public void sauvegarder_modele(String url){
		try {
			FileOutputStream f = new FileOutputStream(url);
			ObjectOutputStream out = new ObjectOutputStream(f);
			out.writeObject(modele);
			out.close();
			f.close();
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public void charger_evaluation(String url){
		try {
			BufferedReader f = new BufferedReader(new InputStreamReader(new FileInputStream(url), "UTF-8"));
			String l;
			while((l = f.readLine()) != null) {
				String[] info = l.toLowerCase().split("	");
				if (info.length > 1) this.eval.add(new Tuple<String, String>(info[0], info[1]));
			}
		} catch(IOException e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public double evaluer(int n, int m){//Mean reciprocal rank
		if (n == -1) n = this.eval.size();
		else Collections.shuffle(this.eval);
		double score = 0.0;
		for (int j=0; j<n; j++){
			Tuple<String, String> test = eval.get(j);
			System.out.println("p(" + test.v + "|" + test.k + ")");
			List<Tuple<String, Double>> res = estimer(test.k, m);
			System.out.println("found: " + res);
			double i = res.stream().map(e -> e.k).collect(Collectors.toList()).indexOf(test.v) + 1;
			if (i > 0) score += 1/i;
		}
		score /= n;
		System.out.println("Score = " + score);
		return score;
  }

	public static void main(String[] args) {
		Autocompletion program = new Autocompletion();
		program.entrainer("data/algerie_train.txt");
		program.sauvegarder_modele("./t_java.mdl");
// 		String phrase = "L' Algérie faisant partie du";
// 		List<Tuple<String, Double>> res = program.estimer(phrase, 10);
// 		System.out.println(res);
		program.charger_evaluation("data/algerie_test.txt");
// 		80 exemplaires ; 10 résultats possibles
		double mrr = program.evaluer(80, 10);
		//System.out.println(mrr);
	}
}

class LaplaceBiGram implements Serializable {
	private Map<String, Integer> uniGrams = new HashMap<>();
	private Map<String, Integer> biGrams = new HashMap<>();

	// TODO : compléter cette méthode
	public void entrainer(List<String[]> data){
		uniGrams.put("<s>", 0);
	}//entrainer

	// TODO : compléter cette méthode
	public double noter(String past, String current){
		return  0.0;
	}

	// TODO : compléter cette méthode
	public List<Tuple<String, Double>> estimer(String[] mots){
		List<Tuple<String, Double>> motsScores = new ArrayList<>();
		// ajouter des tuples (mot, score) à cette liste

		Collections.sort(motsScores, new Comparator<Tuple<String, Double>>() {
			@Override
			public int compare(Tuple<String, Double> o1, Tuple<String, Double> o2) {
				double sub = o2.v - o1.v;
				return (int) Math.signum(sub);
			}
		});
		return motsScores;
  }

}

class Tuple<K, V> {
	public final K k;
	public final V v;
	public Tuple(K k, V v) {
		this.k = k;
		this.v = v;
	}

	public String toString(){
		return "(" + k + ", " + v + ")";
	}
}

File: TP/CH03/autocompletion/python/autocompletion.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import math
import json
import random

class LaplaceBiGram:

    def __init__(self):
        self.uni_grams = {'<s>': 0}
        self.bi_grams = {}

    # TODO : compléter cette méthode
    def entrainer(self, data):
        pass

    # TODO : compléter cette méthode
    def noter(self, past, current):
        return 0.0

    # TODO : compléter cette méthode
    def estimer(self, mots):
        mots_scores = []
        # ajouter des tuples (mot, score) à cette liste
        return sorted(mots_scores, key=lambda tab: tab[1], reverse=True)

    def exporter_json(self):
        return self.__dict__.copy()

    def importer_json(self, data):
        for cle in data:
            self.__dict__[cle] = data[cle]

class Autocompletion():
    def __init__(self):
        self.modele = LaplaceBiGram()
        self.eval = []

    def entrainer(self, url):
        f = open(url, 'r')
        data = []
        for l in f: # la lecture ligne par ligne
            phrase = l.strip().lower().split()
            if len(phrase) > 0:
                data.append(phrase)
        f.close()
        self.modele.entrainer(data)

    def estimer(self, phrase, nbr):
        mots_scores = self.modele.estimer(phrase.strip().lower().split())
        return mots_scores[:nbr]


    def charger_modele(self, url):
        f = open(url, 'r')
        data = json.load(f)
        self.modele.importer_json(data)
        f.close()

    def sauvegarder_modele(self, url):
        f = open(url, 'w')
        json.dump(self.modele.exporter_json(), f)
        f.close()

    def charger_evaluation(self, url):
        f = open(url, 'r')
        for l in f: # la lecture ligne par ligne
            info = l.strip().lower().split("	")
            if len(info) < 2 :
                continue
            self.eval.append(info)

    def evaluer(self, n, m): #Mean reciprocal rank
        if n == -1:
            S = self.eval
            n = len(S)
        else :
            S = random.sample(self.eval, n)
        score = 0.0
        for i in range(n):
            test = S[i]
            print('p(', test[1], '|',test[0], ')')
            res = self.estimer(test[0], m)
            print('found:', res)
            words = [e[0] for e in res]
            try:
                i = words.index(test[1]) + 1
                score += 1/i
            except ValueError:
                pass
        score = score/n
        print('Score = ', score)
        return score

if __name__ == '__main__':
    program = Autocompletion()
    program.entrainer('data/algerie_train.txt')
    program.sauvegarder_modele('./tp.json')
    #phrase = "L' Algérie faisant partie du"
    #res = program.estimer(phrase, 10)
    #print(res)
    program.charger_evaluation('data/algerie_test.txt')
    # 80 exemplaires ; 10 résultats possibles
    mrr = program.evaluer(80, 10)
